### ISSUE TITLE
[Feature Request] Pythonic search space in Optuna sweeper plugin

### ISSUE THREAD

**Issue by MattiasDC** [facebookresearch/hydra#1906]
# 🚀 Feature Request

The Optuna sweeper plugin allows defining a search space and supports three different types: int, float and categorical. It would be nice if there exists a fourth one, 'custom'. This type allows more advanced search space definitions, while still keeping simple search space parameters in the existing format.

## Motivation

**Is your feature request related to a problem? Please describe.**

If you want to find the best neural network configuration using hyperparameter optimization, you can consider using dropout or not. If dropout is selected to be used by Optuna, you also need to define the dropout ratio. This describes a conditional extra hyperparameter.

In the current application you could try something like this:
``` {yaml}
search_space:
  use_dropout:
    type: categorical
    choices:
      - True
      - False
  dropout_ratio:
    type: float
    low: 0.2
    high: 0.7
```

This solution is however not optimal as dropout_ratio is not used if use_dropout is false. This makes it unnecessarily harder for the optimization framework to do its work.

A similar but even more difficult problem occurs when you want to parametrize on the number of layers and you have to define the number of inputs for each layer. An example of this problem is defined in Optuna [here](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/002_configurations.html#defining-parameter-spaces). I currently find no way on how to model this in the Optuna plugin.

## Pitch

**Describe the solution you'd like**

``` {yaml}
nr_layers:
  type: int
  low: 2
  high: 6
layer_sizes:
  type: custom
  method: path.file.suggest_layer_sizes
```

In this case _suggest_layer_sizes_ would be a method in the _file_ in the _path_ with the following implementation:
``` {python}
def suggest_layer_sizes(trial):
  nr_layers = trial.params['nr_layers']
  return [ trial.suggest_int(f"layer_size_{i}", 128, 1028) for i in range(nr_layers) ]
```
I'm not sure whether the method should return the list, or should not return anything.

**Describe alternatives you've considered**
An alternative would be to extend the Optuna types with a 'list' type. This would however not solve the dropout case, as conditionals would still not be possible. Also more complex logic would not be possible to model with this approach.

```
nr_layers:
  type: int
  low: 2
  high: 6
layer_sizes:
 type: list
 length: {nr_layers}
 values:
   type: int
   low: 128
   high: 1028
```

**Are you willing to open a pull request?** (See [CONTRIBUTING](../../CONTRIBUTING.md))
YES

## Additional context
I see this feature as a catch-all. Allowing to use the Optuna plugin when some parts of hyperparameter optimization configurations are unsupported.

**Comment by jieru-hu** [comment]
Thanks @MattiasDC for the request. This is an interesting idea. We will evaluate this for Hydra 1.2

**Comment by MattiasDC** [comment]
Hi @jieru-hu,

I gave this some more thought and the proposed solution might not even be flexible enough. Consider _nr_layers_ being defined as above and _layer_sizes_ as well. I could imagine a case where you want to have another _custom_ type that should be based on _layer_sizes_. This implies that we need to be able to annotate dependencies between search space parameters, making configuration more complex.

Having the _suggest_layer_sizes_ method configure multiple parameters on the trial object feels wrong as we annotated it to only configure _layer_sizes_ in the config file.

What do you think about having the _custom_ logic on a level higher, e.g. on the same level as _search_space_? This could be called for example _custom_search_space_ and also has the _method_ attribute. This method would be called as a second-to-last step, after configuring the parameters defined in the search_space, but before the fixed parameters (a.k.a. command line overrides). This way the _custom_search_space_ can use the _search_space_ hyperparameters.

Note that I also found an inefficiency in the current implementation of the Optuna sweeper plugin. The hyperparameters are configured on the _trial_ object and then it is overridden by _fixed_params_. The trial object itself isn't aware about this override. This makes it harder for the Optuna sweeper to optimize in a good way and during reporting it reports the not-used parameter as 'used' with a value that was not used.

If you agree with either of these proposed solutions, can I create a pull request and can you let me know which solution you prefer?

Kind regards,
Mattias

**Comment by jieru-hu** [comment]
hi @MattiasDC thank you for the thoughtful comments! I will take a closer look at your proposal soon. We do have some plans to change how we'd configure all sweeper's search space to make the user experience more consistent. So it is a good time to evaluate this issue. In the mean time, if you can, feel free to join our [zulipchat](https://hydra-framework.zulipchat.com/login/) and find me there. Since it's generally easier to communicate on zulip :)

**Comment by MattiasDC** [comment]
Hi @jieru-hu,

I will contact you on zulip if I have further questions. Note that my previous remark on the efficiency is not correct, as those parameters are deleted from the search space :)

**Comment by jieru-hu** [comment]
Thanks for the issues and PR @MattiasDC  and sorry for the delayed response! This would be a great feature to add to Hydra. However instead of writing a custom function that access trial. I wonder if we can use OmegaConf's resolver to address this:
Something like the following, using the 2 examples from the [doc](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/002_configurations.html#defining-parameter-spaces) linked

To define the search space
```yaml
hydra:
  sweeper:
    search_space:
      num_layer:
        type: int
        low: 5
        high: 10
      n_layers:
        tag: interpolate
        type: int
        low: 1
        high: 100
        function: ${gen_layers:num_layers, .type, .low, .high}
      svc_c:
        tag: interpolate
        type: float
        low: 1e-10
        high: 1e10
        function: ${gen_maybe_svc:use_svc, .type, .low, .high}
```

in the python application, define the resolver funtion
```python3
def gen_layers(n_layers: int, **kwargs) -> Dict[str, Any]:
    layer_params = {}
    for i in range(n_layers):
        layer_params[f"n_layers.n_units_l{i}"] = DistributionConfig(**kwargs)

    return layer_params


def gen_potential_param(use_svc: True, **kwargs) -> Dict[str, Any]:
    if use_svc:
        return {"svc": DistributionConfig(**kwargs)}
    else:
        return {}


OmegaConf.register_new_resolver(
    "gen_layers", lambda n_layers, **kwargs: gen_layers(n_layers, **kwargs)
)

OmegaConf.register_new_resolver(
    "gen_maybe_svc", lambda use_svc, low, high, _type: gen_potential_param(use_svc, low, high, _type)
)

```

The sweeper will then interpolate the nodes after `n_layers` has been  suggested by optuna, the resulting config should look something like

```
num_layer: 1
n_layers:
  n_layer_l1: 100
use_svc: True
svc_c: 0.1
```

Implementing this should be relatively straightforward, but it is currently blocked by https://github.com/facebookresearch/hydra/issues/1758 (which is planned for 1.2 and should be supported soon :) )


cc @Jasha10

**Comment by MattiasDC** [comment]
Hi @jieru-hu ,
This solution seems elegant and makes dependencies more clear. I didn't know about adding new resolvers.
I'm wondering how the solution would look like:
- Is this 'tag' element a new member of DistributionConfig?
- So will the flow be:
1. Configure all non-interpolate parameters on the trial object
2. Do interpolation (how is this done?). Does this change the actual search_space DictConfig elements? (e.g. removing the interpolate tags and replacing with the resulting config)
3. Configure all interpolated parameters on the trial object

**Comment by MattiasDC** [comment]
Also note that in #1909 I had to adapt the override syntax 'a=1' to '++a=1' as 'a' isn't always defined yet (e.g. 'n_layerl1' isn't known to the launcher its configuration).

**Comment by Jasha10** [comment]
I just had a meeting with Jieru about this; we concluded that the current implementation in #1909 is advantageous because it lets the user access the `trials` object. For now, let's stick with this `custom_search_space` idea instead of adding a new "interpolate" tag as suggested previously.

**Comment by elidub** [comment]
Thank you for working on this feature! I have looked at #1909 and it's [documentation](https://hydra.cc/docs/1.2/plugins/optuna_sweeper/#experimental--custom-search-space-optimization), but I can't make it work with the dropout usecase as illustrated in the first message of this thread.

### Question
It would be nice if one could setup all the ranges at one central place (the Hydra config), and set the logic in the `custom_search_space`.

Consider the minimal working example:

`config.yaml`

```yaml
defaults:
  - override hydra/sweeper: optuna

hydra:
  sweeper:
    sampler:
      seed: 123
    direction: minimize
    study_name: custom-search-space
    storage: null
    n_trials: 20
    n_jobs: 1

    params:
      use_dropout: True, False
      dropout: interval(0.0, 0.5) # dropout should not be considered as a hparam when use_dropout is False
    custom_search_space: run.configure

use_dropout: False
dropout: 0.
```

`run.py`

```python
import hydra
from omegaconf import DictConfig
from optuna.trial import Trial

@hydra.main(config_path="", config_name="config", version_base=None)
def main(cfg: DictConfig) -> float:
    print(f"{cfg =          }")

    loss: float = 1.

    if cfg.use_dropout:
        # simulate that using dropout DECREASES the performance
        loss += cfg.dropout * 10.

    print('Done!')
    return loss

def configure(cfg: DictConfig, trial: Trial) -> None:

    use_dropout: bool = cfg.use_dropout

    if not use_dropout:
        # How do I remove (or set to zero) the dropout parameter?
        trial.suggest_float("dropout", 0.0, 0.0) # No effect? 

        # Other things I've tried:
        # trial.params['dropout'] = 0.0 # No effect?
        # trial.suggest_categorical("dropout", [0.0]) # Can't change distribution.

    print(f"{trial.params = }")


if __name__ == '__main__':
    main()
```

The print statements are merely for debugging.

How do I disable dropout from the trial if use_dropout = False?

If the above is not possible, then an alternative might the the following.

### Alternative question

It might be possible to set the interval in configure() instead of in config.yaml. However, that also doesn't seem to work.

`config.yaml` the same as above.

`run.py`

```python
import hydra
from omegaconf import DictConfig
from optuna.trial import Trial

@hydra.main(config_path="", config_name="config", version_base=None)
def main(cfg: DictConfig) -> float:
    print(f"{cfg =          }")

    loss: float = 100.

    if cfg.use_dropout:
        # simulate that using dropout INCREASES the performance
        loss -= cfg.dropout * 10.
        
    print('Done!')
    return loss

def configure(cfg: DictConfig, trial: Trial) -> None:

    use_dropout: bool = cfg.use_dropout

    if use_dropout:
        # Is not working. The dropout parameter from config.yaml is not being replaced.
        trial.suggest_float("dropout", 1.0, 5.0)

    print(f"{trial.params = }")


if __name__ == '__main__':
    main()
```

Note that in this example the dropout increases the performance, while in the previous question it was decreasing.

However, the dropout range from `config.yaml` is not being replaced when `use_dropout = True`. Even when I remove/comment the dropout: interval(0.0, 0.5) line from `config.yaml`, it stays the default value of dropout: 0..

If I remove `dropout: interval(0.0, 0.5)` from `config.yaml` and remove the if statement `if use_dropout`, the range is being replaced by `trial.suggest_float("dropout", 1.0, 5.0)`, but that would remove the necessary logic...

Any ideas or suggestions how to solve this?

**Comment by elidub** [comment]
Thank you for working on this feature! I have looked at #1909 and it's [documentation](https://hydra.cc/docs/1.2/plugins/optuna_sweeper/#experimental--custom-search-space-optimization), but I can't make it work with the dropout usecase as illustrated in the first message of this thread.

### Question

It would be nice if one could setup all the ranges at one central place (the Hydra config), and set the logic in the `custom_search_space`. 

Consider the minimal working example:

`config.yaml`

```yaml
defaults:
  - override hydra/sweeper: optuna

hydra:
  sweeper:
    sampler:
      seed: 123
    direction: minimize
    study_name: custom-search-space
    storage: null
    n_trials: 20
    n_jobs: 1

    params:
      use_dropout: True, False
      dropout: interval(0.0, 0.5) # dropout should not be considered as a hparam when use_dropout is False
    custom_search_space: run.configure

use_dropout: False
dropout: 0.
```

`run.py`

```python
import hydra
from omegaconf import DictConfig
from optuna.trial import Trial

@hydra.main(config_path="", config_name="config", version_base=None)
def main(cfg: DictConfig) -> float:
    print(f"{cfg =          }")

    loss: float = 1.

    if cfg.use_dropout:
        # simulate that using dropout DECREASES the performance
        loss += cfg.dropout * 10.

    print('Done!')
    return loss

def configure(cfg: DictConfig, trial: Trial) -> None:

    use_dropout: bool = cfg.use_dropout

    if not use_dropout:
        # How do I remove (or set to zero) the dropout parameter?
        trial.suggest_float("dropout", 0.0, 0.0) # No effect? 

        # Other things I've tried:
        # trial.params['dropout'] = 0.0 # No effect?
        # trial.suggest_categorical("dropout", [0.0]) # Can't change distribution.

    print(f"{trial.params = }")


if __name__ == '__main__':
    main()
```

The print statements are merely for debugging.

How do I disable `dropout` from the trial if `use_dropout = False`?

If the above is not possible, then an alternative might the the following.

### Alternative question

It might be possible to set the interval in `configure()` instead of in `config.yaml`. However, that also doesn't seem to work.

`config.yaml` the same as above.

`run.py`

```python
import hydra
from omegaconf import DictConfig
from optuna.trial import Trial

@hydra.main(config_path="", config_name="config", version_base=None)
def main(cfg: DictConfig) -> float:
    print(f"{cfg =          }")

    loss: float = 100.

    if cfg.use_dropout:
        # simulate that using dropout INCREASES the performance
        loss -= cfg.dropout * 10.
        
    print('Done!')
    return loss

def configure(cfg: DictConfig, trial: Trial) -> None:

    use_dropout: bool = cfg.use_dropout

    if use_dropout:
        # Is not working. The dropout parameter from config.yaml is not being replaced.
        trial.suggest_float("dropout", 1.0, 5.0)

    print(f"{trial.params = }")


if __name__ == '__main__':
    main()
```

Note that in this example the dropout increases the performance, while in the previous question it was decreasing.

However, the dropout range from `config.yaml` is not being replaced when `use_dropout = True`. Even when I remove/comment the `dropout: interval(0.0, 0.5)` line from `config.yaml`, it stays the default value of `dropout: 0.`.

If I remove `dropout: interval(0.0, 0.5)` from `config.yaml` and remove the if statement `if use_dropout`, the range is being replaced by `trial.suggest_float("dropout", 1.0, 5.0)`, but that would remove the necessary logic...

Any ideas or suggestions how to solve this?

**Comment by MattiasDC** [comment]
Hi @elidub,

I've just added some information which should get you started on the other comment you made in https://github.com/ashleve/lightning-hydra-template/issues/457#issuecomment-2452973431

**Comment by elidub** [comment]
@MattiasDC Thank you for reply.

I've meant to post the question in this thread, and realized too late that I've asked it accidentally in the other thread. Sorry for cluttering by asking the question twice.
